examples/querying_entity_traits.ipynb

```diff
@@ -0,0 +1,710 @@
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "data = TraitsData()\n",
+    "LocateableContentTrait(data).setLocation(\"file:///some/path\")\n",
+    "BearerTokenTrait(data).setToken(\"==ZxErn43G\")\n",
+    "\n",
+    "entity_ref_or_error = manager.preflight(\n",
+    "    logo_ref, data, PublishAccess.kWrite, context, BatchElementErrorPolicyTag.kVariant)\n",
+    "\n",
+    "helpers.display_result(entity_ref_or_error)"
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "77a3047239a3c76a"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "We get an `InvalidTraitSet` error if the provided trait set is not compatible, or sufficiently complete, for publishing to the entity reference.\n",
+    "\n",
+    "TODO(DF): There is also `InvalidPreflightHint`, but this doesn't obviously apply to `register`. We could use that instead, or rename it, or come up with a new `BatchElementError` code."
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "6c0963e48e593185"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## Errors when querying the trait set\n",
+    "\n",
+    "### Entities that don't exist (yet)\n",
+    "\n",
+    "For `kRead` access, the entity must exist:"
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "71f2b1f16e08b7ae"
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "future_ref = manager.createEntityReference(\"bal:///project_artwork/logos?new\")\n",
+    "\n",
+    "trait_set_or_error = manager.entityTraits(\n",
+    "    logo_ref, EntityTraitAccess.kRead, context, BatchElementErrorPolicyTag.kVariant)\n",
+    "\n",
+    "helpers.display_result(trait_set_or_error)"
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "32f76d73bd99a340"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "If the entity doesn't exist we get an `EntityAccessError`.\n",
+    "\n",
+    "On the other hand, this entity reference is fine for `kWrite`:"
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "73b6ce44255392eb"
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "trait_set_or_error = manager.entityTraits(\n",
+    "    future_ref, EntityTraitAccess.kWrite, context, BatchElementErrorPolicyTag.kVariant)\n",
+    "\n",
+    "helpers.display_result(trait_set_or_error)"
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "1debd3dfd60cb52d"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "Here, once again, we get the minimal trait set required for publishing to this entity reference."
+   ],
+   "metadata": {
+    "collapsed": false
+   },
+   "id": "50f192ae83209477"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "### Read-only entities\n",
+    "\n",
+    "BAL doesn't support publishing to a specific version tag (i.e. doesn't support overwriting a version). \n",
```

Comment:
What do you think to trying to publish to `bal:///project_artwork` instead or similar? I _imagine_ it could be common for "publishing to an old ref to make a new version" (which I think BAL might do already).

Reply:
Yeah this was contrived just for exemplification. IIRC `bal:///project_artwork` would be publishable too, in the current incarnation of BAL. 

I'd be concerned that `bal:///project_artwork` looks like a container, and why wouldn't you be able to publish to that?  Unless we go down the road of `kCreateRelated` vs. `kWrite`, but that's a whole other topic.

Tangential link to the API compliance suite, where we're thinking about the redesign (OpenAssetIO/OpenAssetIO#1191), and have a problem of BAL (and other theoretical managers) not supporting read-only entities, and so must disable those tests.

Reply:
Should fix this:
- OpenAssetIO/OpenAssetIO-Manager-BAL#90

Reply:
Just occurred to me, when you `preflight` in BAL it strips any version identifier from the reference, so there's some justification for saying that you can't publish to an explicit version.

Reply:
It does this to reflect "what will happen" to be helpful, and illustrate how `preflight` can be used. Its pretty essential to allow hosts to publish to an existing ref and so what BAL does. The alternative is some long winded `getRelated` dance that could be entirely avoided, given the manager is in charge of the whole thing anyway.

Reply:
Added an AC to https://github.com/OpenAssetIO/OpenAssetIO-Manager-BAL/issues/90 to address this